### PR TITLE
Port StreamTestDefaultMailbox

### PR DESCRIPTION
**Pull request by cconstantin**
Also included are ports of LoggerMailbox and LoggerMessageQueueSemantics,
plus updates to Mailboxes.GetMailboxType to support 'mailbox-requirement'
and 'mailbox-type' in dispatcher configuration.

### CODE REVIEW

**Comment by Horusiath** [comment]
It's good for me, but maybe @JeffCyr @Aaronontheweb or @rogeralsing have something to add?

**Comment by cconstantin** [comment]
Updated Copyright notice in LoggerMailbox.cs, rest is unchanged.

**Comment by cconstantin** [comment]
@Horusiath @Aaronontheweb most of the changes in this PR are not related to streams, so maybe it's better to send a PR to `dev` with only those changes, and rebase streams? Would avoid merge pain when streams is complete.

File: src/core/Akka/Dispatch/Mailboxes.cs [hunk 1]
```diff
@@ -76,23 +93,53 @@ public Type GetMailboxType(Props props, Config dispatcherConfig)
...
+            var hasMailboxType = dispatcherConfig.HasPath("mailbox-type") &&
+                                 dispatcherConfig.GetString("mailbox-type") != Deploy.NoMailboxGiven;
 
-            if (interfaces.Count > 0)
+            Func<Type, Type> verifyRequirements = mailboxType =>
             {
-                var config = _requirementsMapping[interfaces.First()];
-                var mailbox = config.GetString("mailbox-type");
-                var mailboxType = Type.GetType(mailbox);
+                // TODO Akka code after introducing IProducesMessageQueue
```
Comment (JeffCyr) [inline src/core/Akka/Dispatch/Mailboxes.cs]: Do we want to enforce the requirements anytime soon? If so another issue should be created to complete this.

Comment (cconstantin) [inline src/core/Akka/Dispatch/Mailboxes.cs]: Good point. Opened #1770.

**Comment by Aaronontheweb** [comment]
> most of the changes in this PR are not related to streams, so maybe it's better to send a PR to dev with only those changes

Yes :+1: - good idea

**Pull request closed without merging**